agent/util-scripts/pbench-tool-meister-start

```diff
@@ -732,13 +734,87 @@ def main(_prog: str, cli_params: Namespace) -> int:
+                logger.debug("Host %s reports connection `%s`", host, connection)
+                if connection.status == 0 and connection.stdout:
+                    # The SSH_CONNECTION value is the full `stdout` and we
+                    # don't expect a stderr on success; the format is
+                    #   "origin_node origin_port local_node local_port"
+                    # we only need the origin node because we know that's an
+                    # IP for our local TDS and Redis host that the remote can
+                    # reach.
+                    origin = connection.stdout.split()[0]
+                    origin_ip.add(origin)
+
+                    # Store the origin reported by this remote in its parameter
+                    # block: we'll program this as the connection address when
+                    # we send the init handshake to that remote later.
+                    params["origin_host"] = origin
+                else:
+                    # If `ssh` fails, we can't orchestrate anything on this
+                    # remote, so terminate with an error.
+                    logger.error("Host %s reports %s", host, connection)
+                    return ReturnCode.REMOTENOTREACHABLE
+
+        # Process the collected origin addresses from our remotes. Ideally we
+        # have only a single entry here, since the current BaseServer init
+        # (and the Bottle WSGI server) doesn't support listening on multiple
+        # addresses; however we're going to just take the first returned origin
+        # address and hope they can all connect.
+        #
+        # NOTE: an alternative would be (in the absence of an explicit
+        # connection spec) to always listen on '0.0.0.0', but this eliminates
+        # IPv6, and we currently lack the mechanism to listen on both '0.0.0.0'
+        # and '::' (the IPv6 equivalent, aka '0:0:0:0:0:0:0:0').
+        #
+        # NOTE: If the caller/human supplied explicit PBENCH_REDIS_SERVER or
+        # PBENCH_TOOL_DATA_SINK (--redis-server or --tool-data-sink), we'll use
+        # those, but by default we'll use the ssh connection origin instead of
+        # the local `hostname` which may not be reachable by the remotes (or
+        # necessarily routable at all).
+        if len(origin_ip) != 1:
+            logger.warning(
+                "Remote hosts don't report a single controller "
+                "origin IP: this may be a problem: origins are %s",
+                origin_ip
+            )
+        if len(origin_ip) > 0:
+            logger.debug("Our connection hosts are %s", origin_ip)
+            ip = ipaddress.ip_address(iter(origin_ip).next())
+            if isinstance(ip, ipaddress.IPv6Address):
+                origin = f"[{str(ip)}]"
+            else:
+                origin = str(ip)
+            if not tds_server_spec:
+                tds_server_spec = origin
+            if not redis_server_spec:
+                redis_server_spec = origin
```

Comment:
I misread this code on the first try (and I don't like the "origins are set()" message when the set is empty), so I recommend a slight restructuring:
```
        if len(origin_ip) < 1:
            # This means that all hosts are local, right?  So, there _should_ be
            # nothing to do in this case, because any errors would have caused
            # control to return from this function above.
            pass
        else:
            if len(origin_ip) > 1:
                logger.warning(
                    "Remote hosts don't report a single controller "
                    "origin IP: this may be a problem: origins are %s",
                     origin_ip
                )
            logger.debug("Our connection hosts are %s", origin_ip)
            ip = ipaddress.ip_address(iter(origin_ip).next())
            ...
```
Obviously, the first `if` condition could be reversed to get rid of the `pass` body, but then we would need to put the explanatory comment before it.

Reply:
Again `len(origin_ip) == 0` ideally means we're local; but if not, then the `ssh` remotes didn't get a definition for `SSL_CONNECTION` and are not giving us a clue. I need to separate those two cases and skip the warning only when there are no remotes.